### ISSUE TITLE
[Discussion] Document representation

### ISSUE THREAD

**Issue by ChristophAlt** [ArneBinder/pytorch-ie#120]
I think it is time to take stock before we implement yet another version of our document data structure. The goal of this issue is to collect the use cases and issues we observed so far and develop a proposal for a document representation that satisfies those use cases.

I think we probably agree that a document contains text, should provide means to identify it, and also hold some arbitrary metadata. Hence, the document should have the following attributes:
- `text`
- `doc_id` or `id` (optional)
- `medadata` (optional)

The more difficult part is the annotations, how to manage them, and how to use them in different stages of the training / evaluation / inference cycle.

Let us assume a task that requires named entities, semantic relations, and a document label (e.g. a document topic).

### Training
The dataset defines the complete schema of the document annotations (e.g. entities, relations, label) and the respective dataset loader should produce a dataset, e.g. a list of documents, that follows this schema exactly. In this case, the schema could even be defined as part of the dataset, similar to huggingface datasets' `Feature` definition ([LINK](https://huggingface.co/docs/datasets/master/en/dataset_script#add-dataset-attributes)). Similarly, we need annotation types, e.g.:
- **span:** a span of text, e.g. a word, sentence or paragraph (`Span`)
- **labeled span:** a span of text with a corresponding label, e.g. named entities, part-of-speech-tagged words (`LabeledSpan`)
- **labeled multi-span:** a group of spans with a corresponding label, e.g. for co-reference clusters or linked entities (`LabeledMultiSpan`)
- **binary relation:** a directed label between two arguments (spans), e.g. a semantic relationship (`BinaryRelation`)
- **n-ary relation:** a directed labels between multiple arguments (spans) -- with some argument optional (`NAryRelation`)
- **event:** a multi-argument template of spans, i.e. slots assigned to spans, some optional (with a possible trigger) (`Event`)
- **label:** a document-level label (`Label`)

**Note:** Ideally we are able to support the binary, multi-class, and multi-class multi-label case.

This could be represented as follows:

```python
@dataclass
class TextDocument:
    text: str
	doc_id: Optional[str] = None


class TaskDocument(TextDocument):
    document_label: Label
    named_entities: List[LabeledSpan] = field(default=list)
    semantic_relations: List[BinaryRelation] = field(default=list)
```

Why `TextDocument`? We may want to expand the library to different type of documents in the future, e.g. `ImageDocument`, `HtmlDocument`, `OcrDocument` etc.

After loading the dataset with this schema we would be able to run the whole training cycle.
The problem, however, is that most models and thus most task modules can be applied to different tasks that produce the same annotation type (e.g. a span classifier can be applied to named entity recognition and part-of-speech tagging) and named accordingly (e.g. `named_entities` or `pos_tags`). Hence, in the training stage the task module must be made aware of the annotation name(s) the target(s) should be derived from. This is currently achieved by passing one or multiple  keys (type: str) to the constructor of a task module. There are two possible solutions to this:
- use `getattr(...)` to access an attribute dynamically based on a str
- use `TypedDict` instead of `dataclass`

## Evaluation
Similar to the training stage, the dataset defines the complete schema of the document annotations. In addition this stage requires the concept of a `prediction` , which is similar to an annotation but differs in that it is dynamically added to a document by a model. Predictions are typically used to compare them against the ground truth annotations, to compute more complex evaluation metrics based on the whole document after a post-processing step or after predicting with multiple models (e.g. named entity recognition followed by relation classification). In essence, each annotation also requires that predictions of the same type can be added.

This could be represented by a dataclass, e.g.:

```python
class TaskDocument(TextDocument):
    named_entities: AnnotationList[LabeledSpan]  # AnnotationList[T] signals that there may be multiple annotations of type T
    semantic_relations: AnnotationList[BinaryRelation]
    document_label: Annotation[Label]  # Annotation[T] signals that there is only one annotation of type T
```

A major use case of the evaluation and inference stage is the setting where a model's input depends on the predictions of a preceding model, e.g. a relation classification model uses the spans predicted by a named entity recognition model. There's a discrepancy between training stage, which relies on a documents ground truth annotations (e.g. entities) to create the model input, while in the evaluation (optional) or inference stage (mandatory) ground truth annotations may be unavailable and thus the predictions of a preceding model replace the ground truth annotations. The problem is that the taskmodule is unaware of the stage and thus always assumes that the necessary annotations can be accessed via a document's `.annotation` attribute, which holds only when ground truth annotations are available. The two options are:
- allow the user to specify which attribute should be used to create the necessary model inputs (`.groundtruth` or `.predictions`)
- taskmodule access via a common attribute, e.g. `.annotations`, and transparently point it to `.groundtruth` or `.predictions` depending on the setting

Thus, `Annotation[T]` and `AnnotationList[T]` should have two attributes `.groundtruth` and `.predictions` that can be used to access groundtruth and predicted annotations of type T, either as `T` or `List[T]`.

```python
doc = TaskDocument(...)
...
doc.entities.groundtruth  # List[LabeledSpan] containing the ground truth entity spans, if available, defaults to empty list
doc.entities.predictions  # List[LabeledSpan] containing the predicted entity spans, if added by a model, defaults to empty list
```

One open question is how to best handle the switch between `.groundtruth` and `.predictions` in the taskmodules during evaluation and/or when used in a `Pipeline`.

# Inference
The inference stage is similar to the evaluation stage but it may combine multiple independently trained models into a single pipeline. Note: in this case **pipeline** refers to a set of PyTorch-IE `Pipeline`s applied in parallel or successively (if dependent on the predictions of a preceding model).
